File: backend/services/tmdb_service.py
"""
TMDB Service - The Movie Database API Integration
Provides movie/TV metadata, search, and episode information
"""

import httpx
from typing import Optional, List, Dict, Any
import logging

from ..config import settings

logger = logging.getLogger(__name__)


class TMDBService:
    """Service for interacting with TMDB API"""
    
    def __init__(self):
        self.api_key = settings.tmdb_api_key
        self.base_url = settings.tmdb_base_url
        self.image_base_url = "https://image.tmdb.org/t/p"
    
    def _get_headers(self) -> dict:
        """Get API headers"""
        return {
            "Authorization": f"Bearer {self.api_key}",
            "Content-Type": "application/json"
        }
    
    def _get_params(self, **kwargs) -> dict:
        """Get API params with API key"""
        params = {"api_key": self.api_key}
        params.update(kwargs)
        return params
    
    def _get_poster_url(self, poster_path: Optional[str], size: str = "w500") -> Optional[str]:
        """Convert poster path to full URL"""
        if poster_path:
            return f"{self.image_base_url}/{size}{poster_path}"
        return None
    
    async def search_multi(self, query: str, page: int = 1) -> List[Dict[str, Any]]:
        """Search for movies and TV shows"""
        try:
            async with httpx.AsyncClient() as client:
                response = await client.get(
                    f"{self.base_url}/search/multi",
                    params=self._get_params(query=query, page=page, include_adult=False)
                )
                response.raise_for_status()
                data = response.json()
                
                results = []
                for item in data.get("results", []):
                    if item.get("media_type") in ["movie", "tv"]:
                        results.append(self._format_result(item))
                
                return results
                
        except Exception as e:
            logger.error(f"TMDB search error: {e}")
            return []
    
    async def search_tv(self, query: str, page: int = 1) -> List[Dict[str, Any]]:
        """Search for TV shows only"""
        try:
            async with httpx.AsyncClient() as client:
                response = await client.get(
                    f"{self.base_url}/search/tv",
                    params=self._get_params(query=query, page=page)
                )
                response.raise_for_status()
                data = response.json()
                
                return [self._format_tv_result(item) for item in data.get("results", [])]
                
        except Exception as e:
            logger.error(f"TMDB TV search error: {e}")
            return []
    
    async def search_movie(self, query: str, page: int = 1) -> List[Dict[str, Any]]:
        """Search for movies only"""
        try:
            async with httpx.AsyncClient() as client:
                response = await client.get(
                    f"{self.base_url}/search/movie",
                    params=self._get_params(query=query, page=page)
                )
                response.raise_for_status()
                data = response.json()
                
                return [self._format_movie_result(item) for item in data.get("results", [])]
                
        except Exception as e:
            logger.error(f"TMDB movie search error: {e}")
            return []
    
    async def discover_tv(
        self,
        genres: Optional[List[int]] = None,
        min_rating: float = 7.0,
        page: int = 1,
        sort_by: str = "popularity.desc"
    ) -> List[Dict[str, Any]]:
        """Discover TV shows with filters"""
        try:
            params = self._get_params(
                page=page,
                sort_by=sort_by,
                vote_average_gte=min_rating,
                vote_count_gte=100
            )
            
            if genres:
                params["with_genres"] = ",".join(map(str, genres))
            
            async with httpx.AsyncClient() as client:
                response = await client.get(
                    f"{self.base_url}/discover/tv",
                    params=params
                )
                response.raise_for_status()
                data = response.json()
                
                return [self._format_tv_result(item) for item in data.get("results", [])]
                
        except Exception as e:
            logger.error(f"TMDB discover TV error: {e}")
            return []
    
    async def discover_movies(
        self,
        genres: Optional[List[int]] = None,
        min_rating: float = 7.0,
        page: int = 1,
        sort_by: str = "popularity.desc"
    ) -> List[Dict[str, Any]]:
        """Discover movies with filters"""
        try:
            params = self._get_params(
                page=page,
                sort_by=sort_by,
                vote_average_gte=min_rating,
                vote_count_gte=100
            )
            
            if genres:
                params["with_genres"] = ",".join(map(str, genres))
            
            async with httpx.AsyncClient() as client:
                response = await client.get(
                    f"{self.base_url}/discover/movie",
                    params=params
                )
                response.raise_for_status()
                data = response.json()
                
                return [self._format_movie_result(item) for item in data.get("results", [])]
                
        except Exception as e:
            logger.error(f"TMDB discover movies error: {e}")
            return []
    
    async def get_tv_details(self, tv_id: int) -> Optional[Dict[str, Any]]:
        """Get detailed information about a TV show"""
        try:
            async with httpx.AsyncClient() as client:
                response = await client.get(
                    f"{self.base_url}/tv/{tv_id}",
                    params=self._get_params()
                )
                response.raise_for_status()
                data = response.json()
                
                return {
                    "id": str(data.get("id")),
                    "title": data.get("name"),
                    "type": "series",
                    "overview": data.get("overview"),
                    "poster_url": self._get_poster_url(data.get("poster_path")),
                    "rating": data.get("vote_average"),
                    "first_air_date": data.get("first_air_date"),
                    "genres": [g.get("name") for g in data.get("genres", [])],
                    "number_of_seasons": data.get("number_of_seasons"),
                    "number_of_episodes": data.get("number_of_episodes"),
                    "status": data.get("status"),
                    "networks": [n.get("name") for n in data.get("networks", [])]
                }
                
        except Exception as e:
            logger.error(f"TMDB get TV details error: {e}")
            return None
    
    async def get_episode_details(self, tv_id: int, season: int, episode: int) -> Optional[Dict[str, Any]]:
        """Get details about a specific episode"""
        try:
            async with httpx.AsyncClient() as client:
                response = await client.get(
                    f"{self.base_url}/tv/{tv_id}/season/{season}/episode/{episode}",
                    params=self._get_params()
                )
                response.raise_for_status()
                data = response.json()
                
                return {
                    "season_number": data.get("season_number"),
                    "episode_number": data.get("episode_number"),
                    "name": data.get("name"),
                    "overview": data.get("overview"),
                    "air_date": data.get("air_date"),
                    "runtime": data.get("runtime"),
                    "still_path": self._get_poster_url(data.get("still_path"))
                }
                
        except Exception as e:
            logger.error(f"TMDB get episode details error: {e}")
            return None
    
    async def get_genre_list(self, media_type: str = "tv") -> Dict[int, str]:
        """Get genre ID to name mapping"""
        try:
            async with httpx.AsyncClient() as client:
                response = await client.get(
                    f"{self.base_url}/genre/{media_type}/list",
                    params=self._get_params()
                )
                response.raise_for_status()
                data = response.json()
                
                return {g["id"]: g["name"] for g in data.get("genres", [])}
                
        except Exception as e:
            logger.error(f"TMDB get genres error: {e}")
            return {}
    
    async def get_similar_tv(self, tv_id: int, page: int = 1) -> List[Dict[str, Any]]:
        """Get similar TV shows"""
        try:
            async with httpx.AsyncClient() as client:
                response = await client.get(
                    f"{self.base_url}/tv/{tv_id}/similar",
                    params=self._get_params(page=page)
                )
                response.raise_for_status()
                data = response.json()
                
                return [self._format_tv_result(item) for item in data.get("results", [])]
                
        except Exception as e:
            logger.error(f"TMDB get similar TV error: {e}")
            return []
    
    async def find_by_title(self, title: str) -> Optional[Dict[str, Any]]:
        """Find a show/movie by title (returns first match)"""
        results = await self.search_multi(title)
        return results[0] if results else None
    
    def _format_result(self, item: dict) -> Dict[str, Any]:
        """Format search result for unified output"""
        media_type = item.get("media_type", "tv")
        
        if media_type == "movie":
            return self._format_movie_result(item)
        else:
            return self._format_tv_result(item)
    
    def _format_tv_result(self, item: dict) -> Dict[str, Any]:
        """Format TV show result"""
        year = None
        first_air_date = item.get("first_air_date", "")
        if first_air_date:
            year = int(first_air_date[:4]) if len(first_air_date) >= 4 else None
        
        return {
            "id": str(item.get("id")),
            "title": item.get("name"),
            "type": "series",
            "year": year,
            "rating": item.get("vote_average"),
            "overview": item.get("overview"),
            "poster_url": self._get_poster_url(item.get("poster_path")),
            "genre_ids": item.get("genre_ids", []),
            "popularity": item.get("popularity", 0)
        }
    
    def _format_movie_result(self, item: dict) -> Dict[str, Any]:
        """Format movie result"""
        year = None
        release_date = item.get("release_date", "")
        if release_date:
            year = int(release_date[:4]) if len(release_date) >= 4 else None
        
        return {
            "id": str(item.get("id")),
            "title": item.get("title"),
            "type": "movie",
            "year": year,
            "rating": item.get("vote_average"),
            "overview": item.get("overview"),
            "poster_url": self._get_poster_url(item.get("poster_path")),
            "genre_ids": item.get("genre_ids", []),
            "popularity": item.get("popularity", 0)
        }


# Genre ID mappings (TMDB standard IDs)
TV_GENRE_MAP = {
    "action": 10759,
    "adventure": 10759,
    "animation": 16,
    "comedy": 35,
    "crime": 80,
    "documentary": 99,
    "drama": 18,
    "family": 10751,
    "fantasy": 10765,
    "kids": 10762,
    "mystery": 9648,
    "news": 10763,
    "reality": 10764,
    "romance": 10766,
    "sci-fi": 10765,
    "science fiction": 10765,
    "soap": 10766,
    "talk": 10767,
    "thriller": 10765,
    "war": 10768,
    "western": 37
}

MOVIE_GENRE_MAP = {
    "action": 28,
    "adventure": 12,
    "animation": 16,
    "comedy": 35,
    "crime": 80,
    "documentary": 99,
    "drama": 18,
    "family": 10751,
    "fantasy": 14,
    "history": 36,
    "horror": 27,
    "music": 10402,
    "mystery": 9648,
    "romance": 10749,
    "sci-fi": 878,
    "science fiction": 878,
    "thriller": 53,
    "tv movie": 10770,
    "war": 10752,
    "western": 37
}


# Singleton instance
tmdb_service = TMDBService()

File: backend/src/services/tmdbService.ts
import axios from 'axios';
import { config } from '../config/environment';

const TMDB_BASE_URL = 'https://api.themoviedb.org/3';
const TMDB_IMAGE_BASE = 'https://image.tmdb.org/t/p/w500';

interface TMDBContent {
    id: number;
    title?: string;
    name?: string;
    overview: string;
    poster_path: string | null;
    backdrop_path: string | null;
    vote_average: number;
    release_date?: string;
    first_air_date?: string;
    genre_ids: number[];
    popularity: number;
    media_type?: string;
}

interface TMDBDiscoverResponse {
    page: number;
    results: TMDBContent[];
    total_pages: number;
    total_results: number;
}

// Genre mappings for TMDB
const MOVIE_GENRES: Record<string, number> = {
    'Action': 28,
    'Adventure': 12,
    'Animation': 16,
    'Comedy': 35,
    'Crime': 80,
    'Documentary': 99,
    'Drama': 18,
    'Family': 10751,
    'Fantasy': 14,
    'History': 36,
    'Horror': 27,
    'Music': 10402,
    'Mystery': 9648,
    'Romance': 10749,
    'Sci-Fi': 878,
    'Thriller': 53,
    'War': 10752,
    'Western': 37,
};

const TV_GENRES: Record<string, number> = {
    'Action': 10759,
    'Adventure': 10759,
    'Animation': 16,
    'Comedy': 35,
    'Crime': 80,
    'Documentary': 99,
    'Drama': 18,
    'Family': 10751,
    'Fantasy': 10765,
    'History': 36,
    'Horror': 9648,
    'Music': 10402,
    'Mystery': 9648,
    'Romance': 10749,
    'Sci-Fi': 10765,
    'Thriller': 80,
    'War': 10768,
    'Western': 37,
};

// Reverse mapping helper
const getGenreName = (id: number, type: 'movie' | 'tv'): string => {
    const map = type === 'movie' ? MOVIE_GENRES : TV_GENRES;
    return Object.keys(map).find(key => map[key] === id) || '';
};

export class TMDBService {
    private apiKey: string;

    constructor() {
        this.apiKey = config.tmdbApiKey;
    }

    private async makeRequest<T>(endpoint: string, params: Record<string, string | number> = {}): Promise<T> {
        const url = `${TMDB_BASE_URL}${endpoint}`;
        const response = await axios.get<T>(url, {
            params: {
                api_key: this.apiKey,
                ...params,
            },
        });
        return response.data;
    }

    async discoverContent(options: {
        type: 'movie' | 'tv';
        sortBy: string;
        genre?: string;
        page?: number;
        query?: string;
    }): Promise<{
        results: Array<{
            id: number;
            title: string;
            overview: string;
            posterUrl: string | null;
            backdropUrl: string | null;
            rating: number;
            year: string;
            type: 'movie' | 'series';
            genres: string[];
        }>;
        totalPages: number;
        totalResults: number;
    }> {
        const { type, sortBy, genre, page = 1, query } = options;

        // If there's a search query, use search endpoint
        if (query && query.trim()) {
            return this.searchContent(query, type, page);
        }

        // Map sortBy to TMDB sort parameters
        const sortMapping: Record<string, string> = {
            'Popular': 'popularity.desc',
            'Top Rated': 'vote_average.desc',
            'Top': 'vote_average.desc',
            'Latest': 'primary_release_date.desc',
            'Trending': 'popularity.desc',
        };

        const tmdbSort = sortMapping[sortBy] || 'popularity.desc';

        // Get genre ID
        const genreMap = type === 'movie' ? MOVIE_GENRES : TV_GENRES;
        const genreId = genre && genre !== 'All' ? genreMap[genre] : undefined;

        // Build params
        const params: Record<string, string | number> = {
            sort_by: tmdbSort,
            page,
            include_adult: 'false',
            'vote_count.gte': sortBy === 'Top' || sortBy === 'Top Rated' ? 200 : 0,
        };

        if (genreId) {
            params.with_genres = genreId;
        }

        const endpoint = `/discover/${type}`;
        const data = await this.makeRequest<TMDBDiscoverResponse>(endpoint, params);

        return {
            results: data.results.map((item) => ({
                id: item.id,
                title: type === 'movie' ? item.title || '' : item.name || '',
                overview: item.overview,
                posterUrl: item.poster_path ? `${TMDB_IMAGE_BASE}${item.poster_path}` : null,
                backdropUrl: item.backdrop_path ? `https://image.tmdb.org/t/p/original${item.backdrop_path}` : null,
                rating: Math.round(item.vote_average * 10) / 10,
                year: (type === 'movie' ? item.release_date : item.first_air_date)?.split('-')[0] || '',
                type: type === 'movie' ? 'movie' as const : 'series' as const,
                genres: (item.genre_ids || []).map(id => getGenreName(id, type)).filter(g => g),
            })),
            totalPages: data.total_pages,
            totalResults: data.total_results,
        };
    }

    async searchContent(query: string, type: 'movie' | 'tv', page: number = 1) {
        const endpoint = `/search/${type}`;
        const data = await this.makeRequest<TMDBDiscoverResponse>(endpoint, {
            query,
            page,
            include_adult: 'false',
        });

        return {
            results: data.results
                .filter(item => (item.vote_average || 0) >= 5)
                .map((item) => ({
                    id: item.id,
                    title: type === 'movie' ? item.title || '' : item.name || '',
                    overview: item.overview,
                    posterUrl: item.poster_path ? `${TMDB_IMAGE_BASE}${item.poster_path}` : null,
                    backdropUrl: item.backdrop_path ? `https://image.tmdb.org/t/p/original${item.backdrop_path}` : null,
                    rating: Math.round(item.vote_average * 10) / 10,
                    year: (type === 'movie' ? item.release_date : item.first_air_date)?.split('-')[0] || '',
                    type: type === 'movie' ? 'movie' as const : 'series' as const,
                    genres: (item.genre_ids || []).map(id => getGenreName(id, type)).filter(g => g),
                })),
            totalPages: data.total_pages,
            totalResults: data.total_results,
        };
    }

    async getContentDetails(id: number, type: 'movie' | 'tv') {
        const endpoint = `/${type}/${id}`;
        const data = await this.makeRequest<any>(endpoint, {
            append_to_response: 'credits',
        });

        const directors = type === 'movie'
            ? data.credits?.crew?.filter((c: any) => c.job === 'Director').map((c: any) => c.name) || []
            : data.created_by?.map((c: any) => c.name) || [];

        const cast = data.credits?.cast?.slice(0, 5).map((c: any) => c.name) || [];

        return {
            id: data.id,
            title: type === 'movie' ? data.title : data.name,
            overview: data.overview,
            posterUrl: data.poster_path ? `${TMDB_IMAGE_BASE}${data.poster_path}` : null,
            backdropUrl: data.backdrop_path ? `https://image.tmdb.org/t/p/original${data.backdrop_path}` : null,
            rating: Math.round(data.vote_average * 10) / 10,
            year: (type === 'movie' ? data.release_date : data.first_air_date)?.split('-')[0] || '',
            runtime: type === 'movie' ? data.runtime : data.episode_run_time?.[0] || null,
            genres: data.genres?.map((g: any) => g.name) || [],
            directors,
            cast,
            type: type === 'movie' ? 'movie' as const : 'series' as const,
        };
    }

    async getTrending(mediaType: 'movie' | 'tv' | 'all' = 'all', timeWindow: 'day' | 'week' = 'week') {
        const endpoint = `/trending/${mediaType}/${timeWindow}`;
        const data = await this.makeRequest<TMDBDiscoverResponse>(endpoint);

        return {
            results: data.results.map((item) => {
                const isMovie = item.media_type === 'movie' || (!item.name && item.title);
                const type = isMovie ? 'movie' as const : 'series' as const;
                const tmdbType = isMovie ? 'movie' : 'tv';
                return {
                    id: item.id,
                    title: isMovie ? item.title || '' : item.name || '',
                    overview: item.overview,
                    posterUrl: item.poster_path ? `${TMDB_IMAGE_BASE}${item.poster_path}` : null,
                    backdropUrl: item.backdrop_path ? `https://image.tmdb.org/t/p/original${item.backdrop_path}` : null,
                    rating: Math.round(item.vote_average * 10) / 10,
                    year: (isMovie ? item.release_date : item.first_air_date)?.split('-')[0] || '',
                    type,
                    genres: (item.genre_ids || []).map(id => getGenreName(id, tmdbType)).filter(g => g),
                };
            }),
            totalPages: data.total_pages,
            totalResults: data.total_results,
        };
    }
}

export const tmdbService = new TMDBService();
